redbot/core/settings_caches.py

```diff
@@ -122,157 +122,145 @@ async def set_ignored_guild(self, guild: discord.Guild, set_to: bool):
             ret = self._cached_whitelist[gid].copy()
         else:
             if gid is not None:
-                ret = await self._config.guild_from_id(gid).whitelist()
-                if not ret:
-                    ret = []
+                ret = set(await self._config.guild_from_id(gid).whitelist())
             else:
-                ret = await self._config.whitelist()
+                ret = set(await self._config.whitelist())
 
             self._cached_whitelist[gid] = ret.copy()
 
         return ret
 
-    async def add_to_whitelist(self, guild: Optional[discord.Guild], role_or_user: List[int]):
+    async def add_to_whitelist(self, guild: Optional[discord.Guild], role_or_user: Iterator[int]):
         gid: Optional[int] = guild.id if guild else None
         role_or_user = role_or_user or []
-        if not isinstance(role_or_user, list) or not all(
+        if not isinstance(role_or_user, (set, list)) or not all(
             isinstance(r_or_u, int) for r_or_u in role_or_user
         ):
-            raise TypeError("Whitelisted objects must be a list of ints")
+            raise TypeError("Whitelisted objects must be a list or set of ints")
 
         if gid is None:
             if gid not in self._cached_whitelist:
                 self._cached_whitelist[gid] = await self._config.whitelist()
             for obj_id in role_or_user:
-                if obj_id not in self._cached_whitelist[gid]:
-                    self._cached_whitelist[gid].append(obj_id)
-                    async with self._config.whitelist() as curr_list:
-                        curr_list.append(obj_id)
+                self._cached_whitelist[gid].add(obj_id)
+            await self._config.whitelist.set(list(self._cached_whitelist[gid]))
+
         else:
             if gid not in self._cached_whitelist:
                 self._cached_whitelist[gid] = await self._config.guild_from_id(gid).whitelist()
             for obj_id in role_or_user:
-                if obj_id not in self._cached_whitelist[gid]:
-                    self._cached_whitelist[gid].append(obj_id)
-                    async with self._config.guild_from_id(gid).whitelist() as curr_list:
-                        curr_list.append(obj_id)
+                self._cached_whitelist[gid].add(obj_id)
+            await self._config.guild_from_id(gid).whitelist.set(list(self._cached_whitelist[gid]))
 
     async def clear_whitelist(self, guild: Optional[discord.Guild] = None):
         gid: Optional[int] = guild.id if guild else None
-        self._cached_whitelist[gid] = []
+        self._cached_whitelist[gid] = set()
         if gid is None:
             await self._config.whitelist.clear()
         else:
             await self._config.guild_from_id(gid).whitelist.clear()
 
-    async def remove_from_whitelist(self, guild: Optional[discord.Guild], role_or_user: List[int]):
+    async def remove_from_whitelist(
+        self, guild: Optional[discord.Guild], role_or_user: Iterator[int]
```

Comment:
Wrong type hint (also, look at the other comment below)
```suggestion
        self, guild: Optional[discord.Guild], role_or_user: Iterable[int]
```